### PR TITLE
Migration icbc

### PR DESCRIPTION
**Pull request by quhongwei**
What is the purpose of the change
1. 跨注册中心&注册中心内迁移
2. 应用维度下multiple
3. 元数据高可用

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/migration/MigrationClusterComparator.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+package org.apache.dubbo.rpc.cluster.support.migration;
+
+import org.apache.dubbo.common.extension.SPI;
+import org.apache.dubbo.rpc.Invoker;
+
+import java.util.List;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (zouyx) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/migration/MigrationClusterComparator.java:25]: i think you should delete this duplicate license

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/migration/MigrationClusterInvoker.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ */
+package org.apache.dubbo.rpc.cluster.support.migration;
+
+import org.apache.dubbo.common.URL;
+import org.apache.dubbo.rpc.cluster.ClusterInvoker;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (zouyx) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/migration/MigrationClusterInvoker.java:25]: as above

File: dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/migration/MigrationRuleTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.rpc.cluster.support.migration;
+
+import org.junit.jupiter.api.Test;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (zouyx) [inline dubbo-cluster/src/test/java/org/apache/dubbo/rpc/cluster/support/migration/MigrationRuleTest.java:22]: as abovw

**Review by chickenlj — CHANGES_REQUESTED** [review]

File: dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosServiceDiscovery.java [hunk 4]
```diff
@@ -116,16 +112,10 @@ public void unregister(ServiceInstance serviceInstance) throws RuntimeException
...
-                            NamingEvent event = (NamingEvent) e;
-                            handleEvent(event, listener);
-                        }
-                    });
-                } catch (NacosException e) {
-                    e.printStackTrace();
+            service.subscribe(listener.getServiceNames(), e -> { // Register Nacos EventListener
+                if (e instanceof NamingEvent) {
+                    NamingEvent event = (NamingEvent) e;
+                    handleEvent(event, listener);
```
[the remark this answers is not in the document]
Comment (quhongwei) [reply dubbo-registry/dubbo-registry-nacos/src/main/java/org/apache/dubbo/registry/nacos/NacosServiceDiscovery.java:118]: merge from dubbo 3.0

File: dubbo-common/src/main/java/org/apache/dubbo/config/AbstractInterfaceConfig.java [hunk 5]
```diff
@@ -451,11 +451,18 @@ public void setLayer(String layer) {
...
-        if (application != null) {
+        ApplicationConfig globalApplication = ApplicationModel.getConfigManager().getApplicationOrElseThrow();
+        if (globalApplication == null) {
+            return application;
+        }
+        if (application != null && !application.getName().equals(globalApplication.getName())) {
             return application;
         }
-        return ApplicationModel.getConfigManager().getApplicationOrElseThrow();
+        return globalApplication;
```
[the remark this answers is not in the document]
Comment (quhongwei) [reply dubbo-common/src/main/java/org/apache/dubbo/config/AbstractInterfaceConfig.java:465]: merge from 3.0 ?

**Pull request merged**